route-snapper/src/tests.rs

```diff
@@ -0,0 +1,210 @@
+use crate::*;
+
+// The NodeIDs depend on the real southwark.bin graph! If the path between two nodes happens to
+// include a third node, then a test may be confusing, because it could look like an intermediate
+// point.
+const WAYPT1: Waypoint = Waypoint::Snapped(NodeID(1));
+const WAYPT2: Waypoint = Waypoint::Snapped(NodeID(2));
+const WAYPT3: Waypoint = Waypoint::Snapped(NodeID(3));
+const WAYPT4: Waypoint = Waypoint::Snapped(NodeID(4));
+const WAYPT5: Waypoint = Waypoint::Snapped(NodeID(5));
+
+// TODO Test freehand points with all of the variations of tests below
+
+#[test]
+fn test_route_delete() {
+    let map_bytes = std::fs::read("../examples/southwark.bin").unwrap();
+    let mut snapper = JsRouteSnapper::new(&map_bytes).unwrap();
+
+    // Add a waypoint
+    must_mouseover_waypt(&mut snapper, WAYPT1);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1]);
+
+    // Click again to delete it -- shouldn't work, it's the only one
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1]);
+
+    // Add a second
+    must_mouseover_waypt(&mut snapper, WAYPT2);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1, WAYPT2]);
+
+    // Delete the first waypoint
+    must_mouseover_waypt(&mut snapper, WAYPT1);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT2]);
+}
+
+#[test]
+fn test_route_extend() {
+    let map_bytes = std::fs::read("../examples/southwark.bin").unwrap();
+    let mut snapper = JsRouteSnapper::new(&map_bytes).unwrap();
+    // setRouteConfig is a WASM API awkward to call; just set directly
+    snapper.router.config.extend_route = true;
+
+    // Add many waypoints
+    must_mouseover_waypt(&mut snapper, WAYPT1);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1]);
+
+    must_mouseover_waypt(&mut snapper, WAYPT2);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1, WAYPT2]);
+
+    must_mouseover_waypt(&mut snapper, WAYPT3);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1, WAYPT2, WAYPT3]);
+
+    must_mouseover_waypt(&mut snapper, WAYPT4);
+    snapper.on_click();
+    assert_eq!(
+        snapper.route.waypoints,
+        vec![WAYPT1, WAYPT2, WAYPT3, WAYPT4]
+    );
+
+    // Clicking an existing waypoint will still delete it
+    must_mouseover_waypt(&mut snapper, WAYPT2);
+    snapper.on_click();
+    assert_eq!(snapper.route.waypoints, vec![WAYPT1, WAYPT3, WAYPT4]);
+
+    // TODO Drag
+}
+
+#[test]
+fn test_route_dont_extend() {
+    let map_bytes = std::fs::read("../examples/southwark.bin").unwrap();
+    let mut snapper = JsRouteSnapper::new(&map_bytes).unwrap();
```

Comment:
Was a bit confused by this test. I think we should explicitly set don't extend, who knows, default may change in future. I suggest adding
`snapper.router.config.extend_route = false;`